### DISCUSSION TITLE
Render on server for SEO value but defer hydration/render for specific components on client

### DISCUSSION THREAD

**Discussion by kevinpfox** [vercel/next.js#39398] [Help]
We have hamburger menu drawer that has a ton of expensive HTML content in it. It's not visible to the user on app load so it doesn't need to be hydrated or rendered immediately (from user perspective).

The catch, hamburger menu drawer is main navigation content. For SEO purposes it's required HTML be available for Google bots. Ideally we would like to only render HTML on server but NOT hydrate hamburger menu drawer on the client. We don't want to have to run that JS operation to support the hydration work for a component that's not visible to user (it's only visible on user select). Ideally drawer is dynamic imported / lazy loaded.

Is it possible to have all HTML on initial request, but dynamic import / lazy load drawer when user selects hamburger menu button?

We have done this:
`const [setShowDrawer, showDrawer] = useState(!isClient);`

// On hamburger button select
`showDrawer(true)`

We get this desired behavior. But it feels dirty, since we get the following expected React warning, since drawer HTML comes from server and then it get removed immediately on the client causing an intentional HTML mismatch...
```
Warning: An error occurred during hydration. The server HTML was replaced with client content in <div>. 
See more info here: https://nextjs.org/docs/messages/react-hydration-error
```

What is the correct way to do what we are trying to do so we don't get that React warning? Is it possible?

We would like to employ this strategy (progressive hydration) for below the fold components as well. We only want the in-view portion to have hydration tasks applied to them. Same problem though, all HTML needs to be there for Google bots. And React will complain if there is a mismatch between server and client HTML.

I looked at this
https://github.com/valcol/react-hydration-on-demand
But I would like to know if there is a native Next.js solution to this problem before reaching for a 3rd party library.

Our versions...
Next 12.2.4
React 18.2.0

**Comment by jimdrury** [comment]
hey, this is massively late, but the way i'd handle this would be to just use CSS `display: none` if it's false, instead of having optional HTML.
